libs/@local/hash-graph-types/graph_types/data_type.py

```diff
@@ -96,57 +95,27 @@ async def create_data_type(
         *,
         actor_id: UUID,
         graph: "GraphAPIProtocol",
-    ) -> Annotated[Any, "DataTypeAnnotation"]:
+    ) -> type[RootModel]:
         """Create an annotated type from this schema."""
+        # make sure that we don't duplicate work and create the same model twice
+        if self._cached is not None:
+            return self._cached
+
         # Custom data types will require an actor ID and the graph to be passed in
         _actor_id = actor_id
         _graph = graph
 
         const = self.model_extra.get("const") if self.model_extra else None
 
-        # TODO: Use `Field` instead when multiple fields are supported
-        #   https://github.com/pydantic/pydantic/issues/6349
-        #   https://github.com/pydantic/pydantic/issues/6353
-        class DataTypeAnnotation:
-            @classmethod
-            def __get_pydantic_core_schema__(
-                cls,
-                source_type: Any,  # noqa: ANN401
-                handler: GetCoreSchemaHandler,
-            ) -> CoreSchema:
-                schema = handler(source_type)
-                if const is not None:
-                    return core_schema.no_info_after_validator_function(
-                        cls.validate_const,
-                        schema,
-                    )
-                return schema
-
-            @classmethod
-            def __get_pydantic_json_schema__(
-                cls,
-                schema: CoreSchema,
-                handler: GetJsonSchemaHandler,
-            ) -> JsonSchemaValue:
-                json_schema = handler(schema)
-                json_schema.update(
-                    **{
-                        "$id": self.identifier,
-                        "$schema": self.schema_url,
-                        "title": self.title,
-                        "description": self.description,
-                        "kind": self.kind,
-                    },
-                )
-                if const is not None:
-                    json_schema.update(const=const)
-
-                return json_schema
-
-            @classmethod
-            def validate_const(cls, v: DataType) -> None:
-                if v != const:
-                    msg = f"Value must be {const}"
-                    raise ValueError(msg)
-
-        return Annotated[self._type(), DataTypeAnnotation]
+        type_ = Literal[const] if const is not None else self._type()
+
+        base = type("Base", (DataTypeBase,), {"info": self.type_info()})
+
+        model = create_model(
+            slugify(self.identifier, regex_pattern=r"[^a-z0-9_]+", separator="_"),
+            __base__=(base, RootModel),
+            root=(type_, Field(...)),
+        )
```

Comment:
Is it intended that two models are created now?

Reply:
This simply moved the `create_model` call from `DataTypeReference` to `DataType`